server/rosetta/cosmos/client/client.go

```diff
@@ -0,0 +1,256 @@
+package client
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"strconv"
+
+	"google.golang.org/grpc/metadata"
+
+	"github.com/tendermint/tendermint/rpc/client/http"
+	tmtypes "github.com/tendermint/tendermint/rpc/core/types"
+	"google.golang.org/grpc"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/server/rosetta"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	grpctypes "github.com/cosmos/cosmos-sdk/types/grpc"
+	authclient "github.com/cosmos/cosmos-sdk/x/auth/client"
+	"github.com/cosmos/cosmos-sdk/x/auth/tx"
+	auth "github.com/cosmos/cosmos-sdk/x/auth/types"
+	bank "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+// interface assertion
+var _ rosetta.NodeClient = (*Client)(nil)
+
+const tmWebsocketPath = "/websocket"
+
+// options defines optional settings for SingleClient
+type options struct {
+	interfaceRegistry codectypes.InterfaceRegistry
+	cdc               *codec.ProtoCodec
+}
+
+// newDefaultOptions builds the default options
+func newDefaultOptions() options {
+	cdc, ir := MakeCodec()
+	return options{
+		interfaceRegistry: ir,
+		cdc:               cdc,
+	}
+}
+
+// OptionFunc defines a function that edits single client options
+type OptionFunc func(o *options)
+
+// WithCodec allows to build a client with a predefined interface registry and codec
+func WithCodec(ir codectypes.InterfaceRegistry, cdc *codec.ProtoCodec) OptionFunc {
+	return func(o *options) {
+		o.interfaceRegistry = ir
+		o.cdc = cdc
+	}
+}
+
+// Client implements a single network client to interact with cosmos based chains
+type Client struct {
+	auth auth.QueryClient
+	bank bank.QueryClient
+
+	ir codectypes.InterfaceRegistry
+
+	clientCtx client.Context
+}
+
+// NewSingle instantiates a single network client
+func NewSingle(grpcEndpoint, tendermintEndpoint string, optsFunc ...OptionFunc) (*Client, error) {
```

Comment:
this should return a Rosetta client interface instead of the type. This should make it easier for other apps to extend the client to support their custom grpc query clients too.